### PR TITLE
[CP] Implement AllGather based context parallelism

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132820

Summary:

This implementation does not utilize the benefit that after allgather we can directly perform the SDPA without doing the ring-based SDPA, but we can overlap the communication with the first sharded kv computation. This implementation shows some performance benefit and memory saving compared to the original alltoall implementation in certain cases.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_tensor/experimental/attention.py [hunk 1]
```diff
@@ -30,19 +30,26 @@
 from torch.distributed.tensor.parallel.style import ParallelStyle
 
 
+class _CausalBehavior(Enum):
+    SKIP = None
+    NOT_IS_CAUSAL = False
+    IS_CAUSAL = True
+
+
+class _RotateMethod(Enum):
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:39]: `is_causal` is not an option but a runtime conditional variable. But I agree with this approach. Let me change this.

File: test/distributed/_tensor/test_attention.py [hunk 2]
```diff
@@ -148,7 +152,7 @@ def test_ring_attention_sdpa(
                     cp_out = fn(cp_q, cp_k, cp_v, is_causal=is_causal)
                     cp_out.sum().backward()
 
-                    if not compiled:
+                    if not compiled and rotater == _RotateMethod.ALL_TO_ALL:
```
Comment (wconstab) [inline test/distributed/_tensor/test_attention.py:155]: why do we not check comm counts for allgather?

File: test/distributed/_tensor/test_attention.py [hunk 3]
```diff
@@ -289,6 +294,7 @@ def test_ring_attention_native_transformer(self, is_causal: bool) -> None:
     @with_comms
     @sdpa_kernel(backends=[SDPBackend.FLASH_ATTENTION])
     def test_ring_attention_custom_transformer(self) -> None:
+        _cp_options.rotate_method = _RotateMethod.ALL_TO_ALL
```
Comment (wconstab) [inline test/distributed/_tensor/test_attention.py:297]: should all the CP tests be parametrized and test both rotate methods? or what reason do we prefer to only test A2A in some cases

File: torch/distributed/_tensor/experimental/attention.py [hunk 4]
```diff
@@ -231,6 +238,86 @@ def __call__(
...
+    """
+    def __init__(self, pg: dist.ProcessGroup, seq_dim: int) -> None:
+        self._pg = pg
+        self._seq_dim = seq_dim
+        self._buffer: Optional[torch.Tensor] = None
+        self._idx = 0
+
+    def rotate(self, buffer: torch.Tensor, curr_idx: int) -> torch.Tensor:
+        self._idx = curr_idx
+        if self._buffer is None:
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:286]: this code is actually quite confusing. 

it is strange that we have `self._buffer` and we accept `buffer` arg.  We seem to never assign anything to `self._buffer` and we seem to only use `buffer` but we use `self._buffer` to determine whether to perform the allgather and then we return self._buffer.

it makes me think that somewhere else, we are passing `self._buffer` as input to rotate, but that would be strange!

File: torch/distributed/tensor/experimental/_attention.py [hunk 5]
```diff
@@ -33,6 +33,19 @@
...
+class _CausalBehavior(Enum):
+    SKIP = None
+    NOT_IS_CAUSAL = False
+    IS_CAUSAL = True
+
+
+class _RotateMethod(Enum):
+    ALL_TO_ALL = auto()
+    ALL_GATHER = auto()
+    SEND_RECV = auto()
```
Comment (XilunWu) [inline torch/distributed/tensor/experimental/_attention.py:46]: double confirm: this is not implemented yet, right? If user specifies this rotate method, will it error out?

**Review by XilunWu — APPROVED** [review]
lgtm! Left one question

File: torch/distributed/tensor/experimental/_attention.py [hunk 6]
```diff
@@ -258,6 +266,82 @@ def __call__(
...
+        self._aggregated_buffer = _maybe_wait(self._aggregated_buffer)
+        return self._aggregated_buffer.chunk(dist.get_world_size(self._pg))[idx]
+
+
+def _create_rotater(
+    pg: dist.ProcessGroup, seq_dim: int, method: Optional[_RotateMethod] = None
+) -> _RingRotater:
+    if method is None:
+        method = _cp_options.rotate_method
+    if method == _RotateMethod.ALL_TO_ALL:
```
Comment (XilunWu) [inline torch/distributed/tensor/experimental/_attention.py:337]: though it doesn't make actual difference:
`if method` -> `elif method`

File: torch/distributed/tensor/experimental/_attention.py (hunk 5 above)
Comment (fegin) [reply torch/distributed/tensor/experimental/_attention.py:46]: ye, we should remove it. SEND_RECV is the same as ALL_TO_ALL.

File: torch/distributed/tensor/experimental/_attention.py (hunk 6 above)
Comment (fegin) [reply torch/distributed/tensor/experimental/_attention.py:337]: No, `elif method` will be incorrect. It method is None, then we will use the global value.

**Pull request closed without merging**